src/main/java/com/learning/browsers/Browser.java

```diff
@@ -1,47 +1,58 @@
 package com.learning.browsers;
 
-import com.learning.configuration.PropertiesReader;
-import javax.annotation.PreDestroy;
+import java.util.Arrays;
+import java.util.function.Function;
+import org.openqa.selenium.JavascriptExecutor;
 import org.openqa.selenium.WebDriver;
-import org.openqa.selenium.chrome.ChromeDriver;
-import org.openqa.selenium.edge.EdgeDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.pagefactory.DefaultElementLocatorFactory;
+import org.openqa.selenium.support.ui.WebDriverWait;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
 
-public class Browser {
+@Component
+public class Browser implements DisposableBean {
 
-  private static final String BROWSER_PROPERTIES_NAME = "browser";
-  private static final String PATH_TO_EDGE_DRIVER = "pathToEdgeDriver";
-  private final PropertiesReader propertiesReader;
-  private WebDriver webDriver;
+  private final WebDriver webDriver;
 
-  public Browser(PropertiesReader propertiesReader) {
-    this.propertiesReader = propertiesReader;
-    this.webDriver = generateWebDriver();
+  public Browser(@Autowired WebDriverFactory webDriverFactory) {
+    this.webDriver = webDriverFactory.getWebDriver();
   }
 
-  public WebDriver getWebDriver() {
-    return webDriver;
-  }
-
-  private WebDriver generateWebDriver() {
-    String browserName = propertiesReader.getProperty(BROWSER_PROPERTIES_NAME);
-    if (browserName.equalsIgnoreCase("Chrome")) {
-      return new ChromeDriver();
-    } else if (browserName.equalsIgnoreCase("Edge")) {
-      System.setProperty("webdriver.edge.driver", propertiesReader.getProperty(PATH_TO_EDGE_DRIVER));
-      return new EdgeDriver();
-    } else {
-      throw new BrowserNotSupportedException("Unknown browser.");
-    }
+  public DefaultElementLocatorFactory getElementLocator() {
+    return new DefaultElementLocatorFactory(webDriver);
   }
 
   public String getCurrentPageURL() {
     return webDriver.getCurrentUrl();
   }
 
-  @PreDestroy
-  public void closeBrowser() {
-    webDriver.close();
+  public void open(String websiteURL, Loadable page) {
```

Comment:
I think there is no need to pass `PageBase` or whole page object.
This solution is ok?

Reply:
In most cases this is fine, however, imagine a case where opening a page URL is not enough to open it.
For example, you want to open a sub page, which is not available via URL, because it'a a single page application, then in order to open a page you'll have to: 
# perform login
# navigate to sub page (ie. using some button click)
Passing whole page object, and calling `open()` method on PO moves the knowledge of how to open certain page from test to PO.

Reply:
Yes, you are right, but this can be performed during tests when you want to navigate to specific subpage e.g. : 
`browser.open(page.getUrl())`
`page.performLogin();`
`page.goToSubpage();`
I think current solution do not exclude activities you provided. Also we do not have WebDriver in PageObject so opening page is performed in browser. 
Let's keep it in this way for simplification.